### ISSUE TITLE
Java PKCS11 KeyStore support

### ISSUE THREAD

**Issue by jfallows** [aws/aws-nitro-enclaves-acm#39]
I have been investigating how to support terminating TLS connections at EC2 via Java SSLEngine using ACM public and private certificates. This is much the same use case as described for `nginx` in the [ACM for Nitro Enclaves User Guide](https://docs.aws.amazon.com/enclaves/latest/user/enclaves-user.pdf).

While the `p11ne-agent` service on the parent instance handles generating an `nginx` stanza referencing a `pkcs11` URI for the private key, the corresponding referenced public certificate is copied from `s3` directly to the parent filesystem, bypassing the `pkcs11` machinery. This is entirely reasonable as unlike the key, the certificate does not require protection by the enclave.

Since Java 9, the default keystore format is PKCS12 when stored directly on the filesystem. Java also ships with a built-in PKCS11 provider for the standard KeyStore API. That PKCS11 provider has some specific requirements that would need to be addressed by ACM for Nitro Enclaves to get it working.

https://docs.oracle.com/en/java/javase/16/security/pkcs11-reference-guide1.html#GUID-F068390B-EB41-48A0-A713-B4CBCC72285D

After creating a private certificate in ACM for `*.example.com` and following the instructions in the user guide to verify nginx integration, I attempted to access the same ACM private certificate via Java `keytool` which is a standard utility for interacting with Java keystores.

1. Update `/etc/nitro-enclaves/acm.yaml` to configure `pin` explicitly, e.g. `28d6ca954c7528d6ca954c75`
2. `sudo systemctl restart nitro-enclaves-acm.service`
3. Create `/home/ec2-user/pkcs11.conf`
```
name = p11kit
library = /usr/lib64/libp11-kit.so.0
slotListIndex = 0
```
4. Install Amazon Corretto 16 via [installation guide](https://docs.aws.amazon.com/corretto/latest/corretto-16-ug/generic-linux-install.html)
```
sudo rpm --import https://yum.corretto.aws/corretto.key 
sudo curl -L -o /etc/yum.repos.d/corretto.repo https://yum.corretto.aws/corretto.repo
sudo yum install -y java-16-amazon-corretto-devel
```
5. Update `/usr/lib/jvm/java-16-amazon-corretto/conf/security/java.security` to reference `pkcs11.conf`
```
...
security.provider.12=SunPKCS11 /home/ec2-user/pkcs11.conf
...
```
6. List ACM certificates via `keytool`
```
keytool -J-Djava.security.debug=pkcs11keystore -storetype PKCS11-p11kit -keystore NONE -storepass 28d6ca954c7528d6ca954c75 -list
pkcs11keystore: did not find match for private key with CKA_ID [0x01] (ignoring entry)
Token Alias Map:
  [empty]
pkcs11keystore: P11KeyStore load. Entry count: 0
Keystore type: PKCS11-P11KIT
Keystore provider: SunPKCS11-p11kit

Your keystore contains 0 entries
```

This last result was initially unexpected, but seems to be caused by a mismatch between ACM for Nitro Enclaves and the Java PKCS11 provider requirements linked above.

The Java PKCS11 provider requires both the private key and the trusted certificate to be reachable via `C_FindObjects` on the same `token`. At present, the Java PKCS11 provider correctly discovers the ACM private key, but does not find the matching ACM certificate (chain), so the private key is ignored, leading to the following debug output:
```
pkcs11keystore: did not find match for private key with CKA_ID [0x01] (ignoring entry)
```
According to the Java PKCS11 provider restrictions linked above, the matching ACM certificate would need to have the same `CKA_ID` to be associated with the private key.

If this could be addressed, then it seems likely that TLS termination at EC2 via Java SSLEngine using ACM public and private certificates would just work.

I would be more than happy to help with testing any proposed solutions.

**Comment by alcioa** [comment]
Hi @jfallows , I took a look at the SunPKCS11 Provider KeyStore Requirements requirements you linked. As stated in the docs, in this reference example we used pkcs11 provider. 

As you noticed we currently do not support `CKO_CERTIFICATE` entries in the token when we provision it, this the object iteration shall fail in `C_FindObject` For RSA and EC keys, we only store `CKO_PRIVATE_KEY` and `CKO_PUBLIC_KEY` class objects (i.e. https://github.com/aws/aws-nitro-enclaves-acm/blob/main/src/vtok_p11/src/backend/db/mod.rs#L90). 

Unfortunately, we do not have this planned on the current roadmap, but we gladly accept and review PRs. Or we'll open a feature request when such time will come.

Please let us know how would this work best for you in order to get TLS termination working via the JAVA SSLEngine.

**Comment by alcioa** [comment]
I saw your draft PR. Was wondering if it's the latest version in order to discuss on it.

**Comment by jfallows** [comment]
@alcioa yes, i've just moved that PR to `ready-to-review`, so we can discuss it and iterate as needed.

**Comment by jfallows** [comment]
@alcioa please review linked PR #41 .

**Comment by alcioa** [comment]
Available in v1.1.0 tag and in aws-nitro-enclaves-acm RPM v1.1.0
